Mathlib/Algebra/Ring/SumsOfSquares.lean

```diff
@@ -22,90 +23,150 @@ sum of squares in `R`.
-
-@[deprecated (since := "2024-08-09")] alias isSumSq := IsSumSq
-
-/--
-If `S1` and `S2` are sums of squares in a semiring `R`, then `S1 + S2` is a sum of squares in `R`.
--/
-theorem IsSumSq.add [AddMonoid R] {S1 S2 : R} (p1 : IsSumSq S1)
-    (p2 : IsSumSq S2) : IsSumSq (S1 + S2) := by
-  induction p1 with
-  | zero             => rw [zero_add]; exact p2
+  | sq_add (a S : R) (hS : IsSumSq S) : IsSumSq (a * a + S)
+
+/-- In an additive monoid with multiplication,
+if `S1` and `S2` are sums of squares, then `S1 + S2` is a sum of squares. -/
+theorem IsSumSq.add [AddMonoid R] [Mul R] {S1 S2 : R}
+    (h1 : IsSumSq S1) (h2 : IsSumSq S2) : IsSumSq (S1 + S2) := by
+  induction h1 with
+  | zero             => rw [zero_add]; exact h2
   | sq_add a S pS ih => rw [add_assoc]; exact IsSumSq.sq_add a (S + S2) ih
 
-@[deprecated (since := "2024-08-09")] alias isSumSq.add := IsSumSq.add
+/-- In an additive unital magma with multiplication, `x * x` is a sum of squares for all `x`. -/
+theorem isSumSq_mul_self [AddZeroClass R] [Mul R] (a : R) : IsSumSq (a * a) := by
+  rw [← add_zero (a * a)]
+  exact IsSumSq.sq_add _ _ IsSumSq.zero
 
-variable (R) in
+/-- In an additive unital magma with multiplication `R`, squares in `R` are sums of squares.
+By definition, a square in `R` is a term `x : R` such that `x = y * y` for some `y : R`
+and in Mathlib this is known as `IsSquare R` (see Mathlib.Algebra.Group.Even). -/
+theorem isSumSq_isSquare [AddZeroClass R] [Mul R] {x : R} (hx : IsSquare x) : IsSumSq x := by
+  rcases hx with ⟨a, ha⟩
+  rw [ha]
+  exact isSumSq_mul_self a
+
+/-- Alternative induction scheme for `IsSumSq` using `IsSquare`. -/
+theorem IsSumSq.induction_alt [Mul R] [Add R] [Zero R]
+    {p : (S : R) → (h : IsSumSq S) → Prop} (S : R) (hS : IsSumSq S)
+    (zero : p 0 zero)
+    (sq_add : ∀ x S, (hS : IsSumSq S) → (hx : IsSquare x) → p S hS →
+      p (x + S) (by rcases hx with ⟨y, hy⟩; rw [hy]; exact sq_add y S hS)) : p S hS := by
+  induction hS with
+  | zero => exact zero
+  | sq_add a S hS hS_ih => apply sq_add (a * a) S hS _ hS_ih; use a
+
+/-- Helper lemma for `IsSumSq.mul`. -/
+theorem IsSumSq.mul_isSquare [NonUnitalCommSemiring R] {S x : R}
+    (hS : IsSumSq S) (hx : IsSquare x) : IsSumSq (S * x) := by
+  induction' S, hS using IsSumSq.induction_alt
+  case zero => simpa using zero
+  case sq_add y T hT hy ih => rw [right_distrib]
+                              exact IsSumSq.add (isSumSq_isSquare (IsSquare.mul hy hx)) ih
+
+/-- In a (not necessarily unital) commutative semiring,
+if `S1` and `S2` are sums of squares, then `S1 * S2` is a sum of squares. -/
+theorem IsSumSq.mul [NonUnitalCommSemiring R] {S1 S2 : R}
+    (h1 : IsSumSq S1) (h2 : IsSumSq S2) : IsSumSq (S1 * S2) := by
+  induction' S2, h2 using IsSumSq.induction_alt
+  case zero => simpa using zero
+  case sq_add y T hT hy ih => rw [left_distrib]
+                              exact IsSumSq.add (IsSumSq.mul_isSquare h1 hy) ih
```

Comment:
I don't understand why you are doing induction here. Can't you first prove the obvious theorem that `IsSumSq a` iff `a` is a sum of squares (in terms of `Finset
sum`) then use `Finset.sum_mul_sum` to express the product of sums of squares as a sum of squares.

Reply:
The predicate is defined inductively, so an induction proof is easy. I think the fact that a sum of squares can be written as.. as sum of squares is somehow implicit in the AddSubmonoid closure theorem?

Reply:
As mentioned on Xena, the predicate is defined inductively because it would be annoying to state it using an indexed sum, but that doesn't forbid you from stating the indexed sum. And yes it is somehow implicit in the closure theorem, but you shouldn't expect your users from knowing that when they search for the fact that a sum of squares is a sum of squares!

Reply:
Have found this theorem now (`Submonoid.exists_multiset_of_mem_closure`); still unsure about using `Finset.sum_mul_sum`. In particular, is there a lemma flattening two sums into one?

Reply:
Basically I think `Finset.sum` is really fiddly and so the inducive definition is easier to work with. What I will do though is prove the 'obvious theorem' to characterise `IsSumSq` in terms of `Finset.sum`.

Reply:
> Basically I think `Finset.sum` is really fiddly

That's an issue, not an issme. I can teach you how to use it. It really shouldn't be that hard

Reply:
> As mentioned on Xena, the predicate is defined inductively because it would be annoying to state it using an indexed sum, but that doesn't forbid you from stating the indexed sum. And yes it is somehow implicit in the closure theorem, but you shouldn't expect your users from knowing that when they search for the fact that a sum of squares is a sum of squares!

Hi @YaelDillies ! May I ask what you mean when say 'as mentioned on Xena'?

Reply:
Artie brought up my comment above on Kevin's Xena Discord server, and I replied there

Reply:
Done